synapse/storage/databases/main/events_worker.py

```diff
@@ -586,19 +586,38 @@ def _get_events_from_db(self, event_ids, allow_rejected=False):
             room_version_id = row["room_version_id"]
 
             if not room_version_id:
-                # this should only happen for out-of-band membership events
-                if not internal_metadata.get("out_of_band_membership"):
-                    logger.warning(
-                        "Room %s for event %s is unknown", d["room_id"], event_id
+                # this should only happen for out-of-band membership events which
+                # arrived before #6983 landed. For all other events, we should have
+                # an entry in the 'rooms' table.
+                #
+                # However, the 'out_of_band_membership' flag is unreliable for older
+                # invites, so just accept it for all membership events.
+                #
+                if d["type"] != EventTypes.Member:
```

Comment:
I'd be sorely tempted to do this just for invites (and leaves?). Do we also want to assert that they are outliers?

Reply:
> I'd be sorely tempted to do this just for invites (and leaves?)

considered that, but it would mean going fishing in the depths of the event dict, which felt slightly excessive. Could do it though.

> Do we also want to assert that they are outliers?

possibly. it's worth noting that out-of-band invite rejections weren't flagged as outliers until 0.99.0, which I think could be problematic too, though I guess they are less likely to get loaded in a `/sync`, so... :man_shrugging:

Reply:
Fair enough. I'd lean on being strict where we know is fine, but my feelings on this I would definitely not describe as "strong", so :shrug: